src/swarm-topography/Swarm/Game/Scenario/Topography/Structure/Recognition/Precompute.hs

```diff
@@ -79,85 +81,74 @@ import Swarm.Util (histogram)
 -- a 'Reader'.
 --
 -- 's' is the state variable, 'a' is the return type.
-type GenericEntLocator s a = Cosmic Location -> s (Maybe a)
-
-getEntityGrid :: (Maybe b -> Maybe a) -> NamedGrid (Maybe b) -> [[Maybe a]]
-getEntityGrid extractor = getRows . fmap extractor . structure
+type GenericEntLocator s a = Cosmic Location -> s (AtomicKeySymbol a)
 
 -- | Create Aho-Corasick matchers that will recognize all of the
 -- provided structure definitions
 mkAutomatons ::
   (Ord a, Hashable a) =>
-  (Maybe b -> Maybe a) ->
-  [SymmetryAnnotatedGrid (Maybe b)] ->
-  RecognizerAutomatons (Maybe b) a
-mkAutomatons extractor xs =
-  RecognizerAutomatons
-    infos
-    (mkEntityLookup rotatedGrids)
+  (b -> NonEmptyGrid (AtomicKeySymbol a)) ->
+  [NamedArea b] ->
+  Either RedundantOrientations (RecognizerAutomatons b a)
+mkAutomatons extractor rawGrids = do
+  onlyNonempties <- mapM checkSymmetry extractedItems
+  let rotatedGrids = concatMap (extractGrids . grid) onlyNonempties
+      infos =
+        M.fromList $
+          map ((name . originalItem . grid . annotatedGrid &&& id) . process) onlyNonempties
+  return $
+    RecognizerAutomatons
+      infos
+      (mkEntityLookup rotatedGrids)
  where
-  rotatedGrids = concatMap (extractGrids extractor . namedGrid) xs
+  extractedItems = map (uncurry ExtractedArea . fmap (extractor . structure) . dupe) rawGrids
```

Comment:
Another major change: the "extraction" of `Entities` from `Cells` (which, from the perspective of the recognizer library, is the opaque type parameter `b`) is now only performed once in this single location, and the result is preserved and shared elsewhere it is needed.  The "extraction" function also used to operate piecewise on each cell within a structure.  Now, its contract is to take an entire "shape" of type `b` and produce a `NonEmptyGrid (Maybe a)`.